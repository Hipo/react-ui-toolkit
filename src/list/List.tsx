import "./_list.scss";

import React, {Fragment, Ref} from "react";
import classNames from "classnames";

import {generateListItemKey} from "./util/listUtils";

export interface ListProps<Item = any> {
  items: Item[];
  children: (item: Item, testid: string, index?: number) => JSX.Element;
  listItemKeyGenerator?: (item: Item, testid: string) => string;
  testid?: string;
  role?: string;
  customClassName?: string;
  placeholderProps?: {
    shouldDisplayPlaceholder: boolean;
    placeholder: React.ReactNode;
  };
  emptyStateProps?: {
    shouldDisplayEmptyState: boolean;
    emptyState: React.ReactNode;
  };
  type?: "unordered" | "ordered" | "description";
}

export type ListElementType = Extract<keyof JSX.IntrinsicElements, "ul" | "ol" | "dl">;

function ListComponent<Item extends any>(
  {
    items,
    children,
    customClassName,
    testid,
    listItemKeyGenerator,
    placeholderProps,
    emptyStateProps,
    type = "unordered",
    role
  }: ListProps<Item>,
  ref?: React.ForwardedRef<ListElementType>
) {
  const listClassName = classNames("list", customClassName);
  let ListTypeElement: Extract<keyof JSX.IntrinsicElements, "ul" | "ol" | "dl">;

  switch (type) {
    case "ordered":
      ListTypeElement = "ol";
      break;

    case "description":
      ListTypeElement = "dl";
      break;

    default:
      ListTypeElement = "ul";
      break;
  }

  return (
    // @ts-ignore - ref type is compatible
    <ListTypeElement ref={ref} className={listClassName} role={role} data-testid={testid}>
      {items.map((item: Item, index: number) => {
        const listItemTestId = `${testid}.item-${index}`;

        return (
          <Fragment
            key={generateListItemKey({listItemKeyGenerator, listItemTestId, item})}>
            {children(item, listItemTestId, index)}
          </Fragment>
        );
      })}

      {placeholderProps?.shouldDisplayPlaceholder && placeholderProps.placeholder}

      {!placeholderProps?.shouldDisplayPlaceholder &&
        emptyStateProps?.shouldDisplayEmptyState &&
        emptyStateProps.emptyState}
    </ListTypeElement>
  );
}

const List = React.forwardRef<ListElementType, ListProps>(ListComponent);

export default List as <Item = any>(
  props: ListProps<Item> & {ref?: Ref<ListElementType>}
) => JSX.Element;
