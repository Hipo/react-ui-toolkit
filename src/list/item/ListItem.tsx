import "./_list-item.scss";

import React from "react";
import classNames from "classnames";

import {KEYBOARD_EVENT_KEY} from "../../core/utils/keyboard/keyboardEventConstants";

interface ListItemProps {
  children?: React.ReactNode;
  testid?: string;
  customClassName?: string;
  clickableListItemProps?: {
    onClick: React.ReactEventHandler;
    tabIndex?: number;
  };
  id?: string;
  role?: string;
  ariaSelected?: boolean;
  listItemRef?: React.RefObject<HTMLLIElement>;
}

function ListItem({
  id,
  testid,
  children,
  customClassName,
  clickableListItemProps,
  role,
  ariaSelected,
  listItemRef
}: ListItemProps) {
  const containerClassName = classNames("list-item", customClassName);
  let listItem = (
    <li
      ref={listItemRef}
      id={id}
      data-testid={testid}
      className={containerClassName}
      role={role}
      aria-selected={ariaSelected}>
      {children}
    </li>
  );

  if (clickableListItemProps) {
    listItem = (
      <li ref={listItemRef} id={id} data-testid={testid} className={containerClassName}>
        <div
          role={"button"}
          tabIndex={clickableListItemProps.tabIndex || 0}
          className={"list-item__click-wrapper"}
          onClick={handleClick}
          onKeyPress={handleKeyPress}>
          {children}
        </div>
      </li>
    );
  }

  return listItem;

  function handleClick(event: React.SyntheticEvent) {
    clickableListItemProps!.onClick(event);
  }

  function handleKeyPress(event: React.KeyboardEvent) {
    if (event.key === KEYBOARD_EVENT_KEY.ENTER) {
      handleClick(event);
    }
  }
}

export default ListItem;
