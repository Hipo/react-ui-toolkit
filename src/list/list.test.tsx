import React from "react";
import {render, screen} from "@testing-library/react";
import "@testing-library/jest-dom";
import {create} from "react-test-renderer";

import {testA11y} from "../core/utils/test/testUtils";
import List, {ListProps} from "./List";
import ListItem from "./item/ListItem";

describe("<List />", () => {
  const listItems = ["Hipo", "Labs"];
  const defaultListProps: ListProps = {
    testid: "list",
    items: listItems,
    children: (item, testId, index) => (
      <ListItem key={`list-item-${index}`} testid={testId}>
        <span>{item}</span>
      </ListItem>
    )
  };

  it("should render correctly", () => {
    render(<List {...defaultListProps} />);
  });

  it("should match snapshot", () => {
    const tree = create(<List {...defaultListProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should pass a11y test", async () => {
    const {container} = render(<List {...defaultListProps} />);

    await testA11y(container);
  });

  it("should render children correctly", () => {
    const {getByTestId} = render(<List {...defaultListProps} />);

    expect(getByTestId(defaultListProps.testid!).children).toHaveLength(listItems.length);

    for (let index = 0; index < listItems.length; index++) {
      const listItemContent = getByTestId(`list.item-${index}`);

      expect(getByTestId(defaultListProps.testid!).children[index]).toEqual(
        listItemContent
      );
    }
  });

  it("should render placeholder correctly", () => {
    const {getByTestId} = render(
      <List
        placeholderProps={{
          shouldDisplayPlaceholder: true,
          placeholder: <p data-testid={"list.placeholder"}>{"Loading..."}</p>
        }}
        {...defaultListProps}
      />
    );

    expect(getByTestId(defaultListProps.testid!)).toContainElement(
      getByTestId("list.placeholder")
    );
  });

  it("should render empty state correctly", () => {
    const {getByTestId} = render(
      <List
        emptyStateProps={{
          shouldDisplayEmptyState: true,
          emptyState: <p data-testid={"list.empty-state"}>{"Not found."}</p>
        }}
        {...defaultListProps}
      />
    );

    expect(getByTestId(defaultListProps.testid!)).toContainElement(
      getByTestId("list.empty-state")
    );
  });

  it("should render placeholder if both shouldDisplayPlaceholder and shouldDisplayEmptyState are true", () => {
    const {getByTestId} = render(
      <List
        emptyStateProps={{
          shouldDisplayEmptyState: true,
          emptyState: (
            <p data-testid={"list.empty-state"} className={"list--empty"}>
              {"Not found."}
            </p>
          )
        }}
        placeholderProps={{
          shouldDisplayPlaceholder: true,
          placeholder: <p data-testid={"list.placeholder"}>{"Loading..."}</p>
        }}
        {...defaultListProps}
      />
    );

    expect(getByTestId(defaultListProps.testid!)).toContainElement(
      getByTestId("list.placeholder")
    );
    expect(screen.queryByText("Not found.")).not.toBeInTheDocument();
  });

  it("should handle empty items correctly", () => {
    const {getByTestId} = render(<List {...defaultListProps} items={[]} />);

    expect(getByTestId(defaultListProps.testid!).children).toHaveLength(0);
  });
});
