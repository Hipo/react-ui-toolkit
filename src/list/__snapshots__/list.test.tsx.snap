// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should match snapshot 1`] = `
<ul
  className="list"
  data-testid="list"
>
  <li
    className="list-item"
    data-testid="list.item-0"
  >
    <span>
      Hipo
    </span>
  </li>
  <li
    className="list-item"
    data-testid="list.item-1"
  >
    <span>
      Labs
    </span>
  </li>
</ul>
`;

exports[`<List /> should matches snapshot 1`] = `
<ul
  className="list"
  data-testid="list"
>
  <li
    className="list-item"
    data-testid="list.item-0"
  >
    <span>
      Hipo
    </span>
  </li>
  <li
    className="list-item"
    data-testid="list.item-1"
  >
    <span>
      Labs
    </span>
  </li>
</ul>
`;
